databuilder/extractor/kafka_source_extractor.py

```diff
@@ -0,0 +1,144 @@
+    CONSUMER_GROUP_ID = 'group.id'
+    # We don't deserde the key of the message.
+    # CONSUMER_VALUE_DESERDE = 'value.deserializer'
+    # Each Kafka extractor should only consume one single topic. We could extend to consume more topic if needed.
+    TOPIC_NAME_LIST = 'topic_name_list'
+
+    # Time out config. It will abort from reading the Kafka topic after timeout is reached. Unit is seconds
+    CONSUMER_TOTAL_TIMEOUT = 'consumer_total_timeout'
+
+    # The timeout for consumer polling messages. Default to 1 sec
+    CONSUMER_POLL_TIMEOUT = 'consumer_poll_timeout'
+
+    RAW_VALUE_TRANSFORMER = 'raw_value_transformer'
+
+    def init(self, conf):
+        # type: (ConfigTree) -> None
+        self.conf = conf
+        # pyhocon doesn't allow dot to be in value dict,group.id -> group_id
+        self.consumer_config = conf.get_config(KafkaSourceExtractor.CONSUMER_CONFIG).\
+            as_plain_ordered_dict()
+
+        # convert group_id back to group.id
+        # noted: underscore is never used in Kafka consumer config
+        self.consumer_config = {k.replace('_', '.'): v for k, v in self.consumer_config.items()}
+
+        self.topic_names = conf.get_list(KafkaSourceExtractor.TOPIC_NAME_LIST)  # type: list
+
+        if not self.topic_names:
+            raise Exception('Kafka topic needs to be provided by the user.')
+
+        self.consumer_total_timeout = conf.get_int(KafkaSourceExtractor.CONSUMER_TOTAL_TIMEOUT,
+                                                   default=10)
+
+        self.consumer_poll_timeout = conf.get_int(KafkaSourceExtractor.CONSUMER_POLL_TIMEOUT,
+                                                  default=1)
+
+        # Transform the protoBuf message with a transformer
+        self.val_transformer = conf.get(KafkaSourceExtractor.RAW_VALUE_TRANSFORMER)
+        if self.val_transformer is None:
+            raise Exception('A message transformer should be provided.')
+
+        # Consumer init
+        try:
+            self.consumer = Consumer(self.consumer_config)
+            # TODO: to support only consume a subset of partitions.
+            self.consumer.subscribe(self.topic_names)
+        except Exception:
+            raise RuntimeError('Consumer could not start correctly!')
+
+    def extract(self):
+        # type: () -> Any
+        """
+        :return: Provides a record or None if no more to extract
+        """
+        records = self.consume()
+        for record in records:
+            try:
+                transform_record = self.val_transformer(record)
+                yield transform_record
+            except Exception as e:
+                # Has issues tranform / deserde the record. drop the record
+                LOGGER.exception(e)
+
+    def on_success(self):
+        # Type: () -> None
+        """
+        Commit the offset once get the success callback and close the consumer.
```

Comment:
How about mentioning about registering callback to publisher as one example?
https://github.com/lyft/amundsendatabuilder/blob/master/databuilder/publisher/base_publisher.py#L50